File: 3650-minimum-cost-path-with-edge-reversals/3650-minimum-cost-path-with-edge-reversals.cpp
class Solution {
public:
    int minCost(int n, vector<vector<int>>& edges) {
        vector<vector<pair<int, int>>> graph(n);
        for(int i = 0; i < edges.size(); i++)
        {
            graph[edges[i][0]].push_back({edges[i][1], edges[i][2]});
            graph[edges[i][1]].push_back({edges[i][0], edges[i][2] * 2});
        }
        return dijk(n, graph);
    }
private:
    int dijk(int n, vector<vector<pair<int, int>>>& graph)
    {
        vector<int> res(n, -1);
        priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> pq;
        unordered_set<int> visited;
        pq.push({0, 0});
        while(!pq.empty())
        {
            int dest = pq.top().second;
            int cost = pq.top().first;
            pq.pop();
            if(visited.count(dest))
            {
                continue;
            }
            visited.insert(dest);
            for(int i = 0; i < graph[dest].size(); i++)
            {
                if(res[graph[dest][i].first] == -1 || cost + graph[dest][i].second < res[graph[dest][i].first])
                {
                    res[graph[dest][i].first] = cost + graph[dest][i].second;
                    pq.push({cost + graph[dest][i].second, graph[dest][i].first});
                }
            }
        }
        return res[n - 1];
    }
};

File: 3650-minimum-cost-path-with-edge-reversals/3650-minimum-cost-path-with-edge-reversals.py
class Solution:
    def minCost(self, n: int, edges: List[List[int]]) -> int:
        graph = [[] for _ in range(n)]
        for e in edges:
            graph[e[0]].append((e[1], e[2]))
            graph[e[1]].append((e[0], 2 * e[2]))
        def dijk() -> int:
            res = [-1 for _ in range(n)]
            pq = [(0, 0)]
            visited = set()
            while len(pq) > 0:
                cost, source = heapq.heappop(pq)
                if source in visited:
                    continue
                visited.add(source)
                for dest, travelCost in graph[source]:
                    if res[dest] == -1 or cost + travelCost < res[dest]:
                        res[dest] = cost + travelCost
                        heapq.heappush(pq, (cost + travelCost, dest))
            return res[-1]
        return dijk()